---
title: "Practical - intermediate"
#knit: unilur::knit
author: "Aurelien Ginolhac"
date: "2^nd^ June 2016"
output:
  unilur::tutorial_html:
    solution: no
    theme: cosmo
    credit: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache = TRUE)
```

## Project - set-up

- Create a new project in a meaningful folder name on your computer (such as `R_workshop/day1-intermediate`) using the project manager utility on the upper-right part of the rstudio window.

- Check if you have all those libraries installed
```{r}
library("tidyr")
library("dplyr", warn.conflicts = FALSE)
library("ggplot2")
library("broom")
suppressPackageStartupMessages(library("GEOquery")) # bioconductor is verbose
theme_set(theme_bw(14)) # if you wish to get this theme by default
```

## Aim

Working with GEO datasets could be an hassle and you are going to experience it.
Extensive manipulation of tables (`data.frame` and `matrix`) is required and provides a nice exercise.
Here, we will investigate the relationship between the expression of _ENTPD5_ and mir-182 as it was described by the authors. Even if the data are normalised and should be ready to use, quite an extensive amount of work is still required to reproduce the claimed result.

## Retrieve GEO study

The GEO dataset of interest is [GSE35834](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE35834)

- load the study using the `getGEO` function

```{r}
gse35834 <- getGEO("GSE35834", GSEMatrix = TRUE)
show(gse35834)
```

- what kind of object is `gse35834`?
```{asis, solution=TRUE}
As shown in the Environment tab, it is a list composed by two elements.
Each list is also a list with a special class 'ExpressionSet'.
```

- Two platforms were used in this study, which ones?  
```{asis, solution=TRUE}
according to the GEO webpage:
- GPL15236 ([HuEx-1_0-st] Affymetrix Human Exon 1.0 ST Array)
http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL15236
- GPL8786 ([miRNA-1_0] Affymetrix miRNA Array)
http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL8786
```

- How can you assign the mRNA or mir data to each element of `gse35834`?
```{asis, solution=TRUE}
The function `show()` displays
1. GPL15236
2. GPL8786
Thus, gse35834[[1]] is mRNA (22486 probes)
gse35834[[2]] is mir (7815 probes)
```

## Explore the mRNA expression meta-data

Informations about samples are accessible using `phenoData()` and can directly be retrieved as a `data.frame` with `pData()`.

for example, the following command will return the mRNA meta-data as a `data.frame`
```{r, eval=FALSE}
pData(gse35834[[1]]) 
```

- Extract as a `tbl_df` named `rna_meta` the mRNA meta-data and
    + rename `geo_accession` to `sample`
    + select `source_name_ch1` and all columns that start with "charact" 
    
```{r, solution=TRUE}
rna_meta <- pData(gse35834[[1]]) %>%
  tbl_df() %>%
  select(sample = geo_accession,
         source_name_ch1,
         starts_with("charact"))
```

## Explore the mir expression meta-data

- Extract as a `tbl_df` named `mir_meta` the mRNA meta-data and
    + rename `geo_accession` to `sample`
    + select `source_name_ch1` and all columns that start with "charact" 
    
```{r, solution=TRUE}
mir_meta <- pData(gse35834[[2]]) %>%
  tbl_df() %>%
  select(sample = geo_accession,
         source_name_ch1,
         starts_with("charact"))
```


## Join meta-data

- Explore the two data frames with `View(rna_meta)` and `View(mir_meta)`. Are the samples `GSM*` identical?
```{asis, solution=TRUE}
No, they aren't. This is really annoying as the expression data contain only GSM ids.
```

Then, we would like to somehow join both informations.  
Knowing that both data frames have different "sample" columns, merge them to get the correspondence between RNA `GSM*` and mir `GSM*`. Save the result as `rna_mir`.
```{asis, box = "blue", boxtitle = "Note"}
When 2 data.frames are joined by specific columns and the remaining columns have have
identical names, a '.x' or '.y' suffix is appended for the first and second data frames respectively
```

```{r, solution=TRUE}
inner_join(rna_meta, mir_meta, 
           by = c("characteristics_ch1.1", "characteristics_ch1",
                  "source_name_ch1", "characteristics_ch1.2", 
                  "characteristics_ch1.3", "characteristics_ch1.4", 
                  "characteristics_ch1.5", "characteristics_ch1.6", 
                  "characteristics_ch1.7", "characteristics_ch1.8")) -> rna_mir
```


## Get RNA expression data for the ENTPD5 gene

Expression data can be accessed via `exprs()` which returns a matrix.

```{asis, box = "orange", boxtitle = "Warning"}
If you do not pipe the command to `head`, `R` would print **ALL** rows (or until it reaches `max.print`).
```

```{r, eval=FALSE}
exprs(gse35834[[1]]) %>% head()
```

rows are probes and columns are sample ids in the form `GSM*`.

Probe ids are not meaningful, but `fData()` provides features.
```{r, eval=FALSE}
fData(gse35834[[1]]) %>% head()
```

Again, we need to merge both informations to assign the expression data to the gene of interest.

1. Find the common values that could help us joining.
```{asis, solution=TRUE}
the probe ids are the common values
```
2. A `matrix` contains only numerical values. But, the `rownames` contain the necessary info.  
Transform the `matrix` into a `data.frame`. Then, convert the `rownames` to a column using `tibble::rownames_to_column(var = "ID")`.  
Save as `rna_expression`

```{r, solution=TRUE}
exprs(gse35834[[1]]) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ID") -> rna_expression
```
  
3. merge expression data to platform annotation (`fData(gse35834[[1]])`).
Save as `rna_expression`. `R` is always working on temporary objects, you won't erase the object you are working on.
```{asis, box = "blue", boxtitle = "Note"}
Warnings about *factors being coerced to characters* can be ignored. Factors shouldn't be in the first place (default of `readr` functions)
```
```{r, solution=TRUE}
rna_expression %>%
 inner_join(fData(gse35834[[1]])) -> rna_expression
```

4. Find the [Entrez](http://www.ncbi.nlm.nih.gov/gene) gene id for _ENTPD5_. Usually, the gene symbol is given in the annotation, but each GEO submission is a new discovery.
```{asis, solution=TRUE}
957, for Homo sapiens
```

5. Filter `rna_expression` for the gene of interest and tidy the samples:  
A column `sample` for all `GSM*` and a column `rna_expression` containing the expression values.
Save the result as `rna_expression_melt`. At this point you should get a `data.frame` of 80 values.
```{r, solution=TRUE}
rna_expression %>%
  filter(ENTREZ_GENE_ID == 957) %>% 
  gather(sample, rna_expression, starts_with("GSM")) -> rna_expression_melt
```

6. Add the meta-data and discard the columns `ID`, `SPOT_ID` and `sample.x`. Save the result as `rna_expression_melt`.
```{r, solution=TRUE}
rna_expression_melt %>%
  inner_join(rna_mir, by = c("sample" = "sample.x")) %>%
  select(-ID, -SPOT_ID, -sample.y) -> rna_expression_melt
```

## Get mir expression data for miR-182

1. Repeat the previous step but using `exprs(gse35834[[2]])` for the `mir_expression`.
This time, the mir names are nicely provided by `fData(gse35834[[2]])` in the column `miRNA_ID_LIST`
```{r, solution=TRUE}
exprs(gse35834[[2]]) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ID") %>%
  # match expression data to platform annotation
  inner_join(fData(gse35834[[2]])) %>% 
  gather(sample, mir_expression, starts_with("GSM")) %>% # melt patients
  filter(miRNA_ID_LIST == "hsa-mir-182") -> mir_expression_melt
```

2. How many rows do you obtain? How many are expected?
```{asis, solution=TRUE}
78 samples for the mir experiment, so expect 78, obtain twice this number
```

3. Find out what happened, and plot the boxplot distribution of `expression` by `ID` 
```{asis, solution=TRUE}
The mir array contains probes for both strands of mir:
- mature mir
- immature mir, named "*", star.
```
```{r, solution=TRUE}
mir_expression_melt %>%
  ggplot(aes(x = ID, y = mir_expression))+
  geom_boxplot()
```
```{asis, solution=TRUE}
The immature mir, named star is indeed merely expressed
```

4. Filter out the irrelevant IDs using `grepl` in the `filter` function.
```{asis, box = "lightgreen", boxtitle = "Hint"}
adding `!` to a condition means `NOT`.
Example `filter(iris, !grepl("a", Species))`: remove all Species that contain an "a".
```
```{r, solution=TRUE}
mir_expression_melt %>%
  filter(!grepl("star", ID)) -> mir_expression_melt
```

5. Add the meta-data, count the number of rows. Discard the column `sample.x` after joining.
```{r, solution=TRUE}
mir_expression_melt %>%
  inner_join(rna_mir, by = c("sample" = "sample.y")) %>%
  select(-sample.x) -> mir_expression_melt
```
```{asis, solution=TRUE}
77 rows, we lost GSM875854, which is not present in the meta-data nor the GSE description.
Let it down
```



## join both expression

Join `rna_expression_melt` and `mir_expression_melt` by their common columns EXCEPT `sample`.
Save the result as `expression`.
```{r, solution=TRUE}
expression <- inner_join(rna_expression_melt, mir_expression_melt,
                         by = c("source_name_ch1", "characteristics_ch1", "characteristics_ch1.1", "characteristics_ch1.2", "characteristics_ch1.3", "characteristics_ch1.4", "characteristics_ch1.5", "characteristics_ch1.6", "characteristics_ch1.7", "characteristics_ch1.8"))
```

## Examine gene expression according to meta data

1. Plot the gene expression distribution by Gender. Is there any obvious difference? 
```{r, solution=TRUE}
expression %>%
  ggplot(aes(y = rna_expression, x = characteristics_ch1))+
  geom_boxplot()
```
```{asis, solution=TRUE}
no relation to gender
```

2. Plot gene AND mir expression distribution by Gender. Is there any obvious difference?


```{asis, box = "blue", boxtitle = "Hint"}
You will need to tidy by gathering rna and mir expression
```

```{r, solution=TRUE}
expression %>%
  gather(expression, value, ends_with("expression")) %>%
  ggplot(aes(y = value, x = characteristics_ch1, fill = expression))+
  geom_boxplot()
```
```{asis, solution=TRUE}
no relation to gender for both expressions
```


3. Plot gene AND mir expression distributions by source (control / cancer).
To make it easier, a quick hack is `separate(expression, source_name_ch1, c("source", "rest"), sep = 12)` to get `source` as control / cancer. Is there any difference?
```{r, solution=TRUE}
expression %>%
  gather(expression, value, ends_with("expression")) %>%
  separate(source_name_ch1, c("source", "rest"), sep = 12) %>%
  ggplot(aes(y = value, fill = source, x = expression))+
  geom_boxplot()+ ggtitle("GSE35834")
```
```{asis, solution=TRUE}
Like it is stated in the summary of the study, the expression of mir-182 seems indeed higher in cancer
while the ENTPD5 expression seems lower.
```

4. Replot 3. but reordering the levels so normal colon comes first. 
Display _normal_ in "lightgreen" and _cancer_ in "red" using `scale_fill_manual()`
```{r, solution=TRUE}
expression %>%
  gather(expression, value, ends_with("expression")) %>%
  separate(source_name_ch1, c("source", "rest"), sep = 12)  %>% 
  mutate(source = factor(source, levels = c("normal colon", "colon cancer"))) %>%
  ggplot(aes(y = value, fill = source, x = expression))+
  geom_boxplot()+
  scale_fill_manual(values = c("lightgreen", "red"))
```


## plot relation ENTPD5 ~ mir-182 as scatter-plot for all patients

- add a linear trend using `geom_smooth()` for all data + per source
```{r, solution=TRUE}
expression %>%
  separate(source_name_ch1, c("source", "rest"), sep = 12) %>%
  ggplot(aes(x = mir_expression, y = rna_expression))+
  geom_point(aes(colour = source))+
  geom_smooth(method = "lm", se = FALSE)+
  geom_smooth(aes(colour = source), method = "lm", se = FALSE)+
  labs(y = "gene expression ENTPD5",
       x = "hsa-mir-182 expression")+
  ggtitle("GSE35834")
```

- does it support the claim of the study?
```{asis, solution=TRUE}
the two dot clouds between normal and cancer origin do split by 
- high mir expression / low gene expression
- mild mir expression / high gene expression
but the trend is not so clear
```

## Supplementary exercise - linear regression

- get the estimate from the linear trend.
linear models are outputted by `lm()` as lists. Since `data.frame` are much easier to work with,
[David Robinson](http://varianceexplained.org/) developed [`broom`](https://cran.r-project.org/web/packages/broom/index.html).
We will present the use of `broom` during the advanced lecture, but you can have an insight here and test `broom::tidy()` coupled with `dplyr::do()`.
```{r}
library("broom")
expression %>%
  separate(source_name_ch1, c("source", "rest"), sep = 12) %>%
  group_by(source) %>%
  do(tidy(lm(rna_expression ~ mir_expression, data = .))) %>%
  filter(term != "(Intercept)")
```

The estimate of the intercept is not meaningful thus it is filtered out.
One can easily see that the slope is not significant when data are slipped by source.

- Perform the linear regression and tidy the results for all data, is it significant?
```{r, solution=TRUE}
expression %>%
  do(tidy(lm(rna_expression ~ mir_expression, data = .))) %>%
  filter(term != "(Intercept)")
```
```{asis, solution=TRUE}
with a pvalue of 7.54e-6, the negative is highly significant
```

- replace `tidy` by `glance` to extract the $r^2$. Is this value satisfactory?
```{r, solution=TRUE}
expression %>%
  do(glance(lm(rna_expression ~ mir_expression, data = .)))
```
```{asis, solution=TRUE}
with a r^2 of 0.236, i.e only 23.6% of the variance explained, a linear fit sounds bad due to outliers
```

## Perform a linear model for the expression of _ENTPD5_ and ALL mirs

- Count how many `hsa-mir`, which are not star, are present on the array GPL8786
```{r, solution=TRUE}
fData(gse35834[[2]]) %>%
  filter(grepl("^hsa", ID)) %>%
  filter(!grepl("star", ID)) %>%
  nrow()
```

- Retrieve the expression values for the 677 human mir like you did before.
Same procedure, except that you don't filter for mir-182. Save as `all_mir_rna_expression`
```{r, solution=TRUE}
exprs(gse35834[[2]]) %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "ID") %>%
  filter(grepl("^hsa", ID)) %>%
  # match expression data to platform annotation
  gather(sample, mir_expression, starts_with("GSM")) %>%
  filter(!grepl("star", ID)) %>%
  inner_join(fData(gse35834[[2]])) %>% 
  inner_join(rna_mir, by = c("sample" = "sample.y")) %>%
  select(-sample.x) %>% 
  inner_join(rna_expression_melt,
             by = c("source_name_ch1", "characteristics_ch1", 
                    "characteristics_ch1.1", "characteristics_ch1.2",
                    "characteristics_ch1.3",  "characteristics_ch1.4",
                    "characteristics_ch1.5", "characteristics_ch1.6", 
                    "characteristics_ch1.7", "characteristics_ch1.8")) -> all_mir_rna_expression 
```

- Perform the 677 linear models, tidy the results and arrange by the `adj.r.squared`
```{r, solution=TRUE}
all_mir_rna_expression %>%
  group_by(ID) %>%
   do(glance(lm(rna_expression ~ mir_expression, data = .))) %>%
  ungroup() %>%
  arrange(desc(adj.r.squared))
```

- Get the top 12 mir and plot the scatter plot
```{r, solution=TRUE}
top12_mir <- all_mir_rna_expression %>%
  group_by(ID) %>%
   do(glance(lm(rna_expression ~ mir_expression, data = .))) %>%
  ungroup() %>%
  arrange(desc(adj.r.squared)) %>%
  head(12) %>%
  .$ID

all_mir_rna_expression %>%
  filter(ID %in% top12_mir) %>%
  separate(source_name_ch1, c("source", "rest"), sep = 12) %>%
  ggplot(aes(x = mir_expression, y = rna_expression))+
  geom_point(aes(colour = source))+
  geom_smooth(method = "lm", se = FALSE)+
  facet_wrap(~ ID, ncol = 4)+
  labs(y = "gene expression ENTPD5",
       x = "hsa-mir expression")
```


